File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PropertiesAccessCorrectField.NetFramework.cs
// https://github.com/SonarSource/sonar-dotnet/issues/3442
// Also see "Add WinForms support in unit tests when targeting .Net Core" https://github.com/SonarSource/sonar-dotnet/issues/3426
namespace TestCases
{
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Media;

    public class CustomControl : UserControl
    {
        public static readonly DependencyProperty BackgroundColorProperty =
            DependencyProperty.Register("BackgroundColor", typeof(Color), typeof(CustomControl), new PropertyMetadata(Colors.Red, ColorPropertyChanged));

        public static readonly DependencyProperty ForegroundColorProperty =
            DependencyProperty.Register("ForegroundColor", typeof(Color), typeof(CustomControl), new PropertyMetadata(Colors.Red, ColorPropertyChanged));

        private Color backgroundColor;
        private Color foregroundColor;

        public Color BackgroundColor
        {
            get => (Color)GetValue(BackgroundColorProperty); // Noncompliant - FP
            set => this.SetValue(BackgroundColorProperty, value); // Noncompliant - FP
        }

        public Color ForegroundColor
        {
            get
            {
                return (Color) GetValue(ForegroundColorProperty); // Noncompliant - FP
            }

            set // Noncompliant - FP
            {
                this.SetValue(ForegroundColorProperty, value);
            }
        }

        private static void ColorPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        {
        }
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/PropertiesAccessCorrectField.NetFramework.vb
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Media

Public Class Repro3442
    Inherits  UserControl

    Public Shared ReadOnly BackgroundColorProperty As DependencyProperty = DependencyProperty.Register("BackgroundColor", GetType(Color), GetType(Color), New PropertyMetadata(""))
    Public Shared ReadOnly ForegroundColorProperty As DependencyProperty = DependencyProperty.Register("ForegroundColor", GetType(Color), GetType(Color), New PropertyMetadata(""))
    Public Shared ReadOnly OldColorProperty As DependencyProperty = DependencyProperty.Register("OldColor", GetType(Color), GetType(Color), New PropertyMetadata(""))

    Private backgroundColor_ As Color
    Private foregroundColor_ As Color
    Private oldColor_ As Color

    Public Property BackgroundColor As Color
        Get
            Return GetValue(BackgroundColorProperty)
        End Get

        Set
            SetValue(BackgroundColorProperty, value)
        End Set
    End Property

    Public Property ForegroundColor As Color
        Get
            Return GetValue(ForegroundColorProperty)
        End Get

        Set
            SetValue(ForegroundColorProperty, value)
        End Set
    End Property

    Public Property OldColor As Color
        Get
            Return GETVALUE(OldColorProperty)
        End Get

        Set
            SETVALUE(OldColorProperty, value)
        End Set
    End Property

End Class
